### ISSUE TITLE
false positive on chained re-exports of ffi C types with `disallowed-types` lint

### ISSUE THREAD

**Issue by fidelicura** [rust-lang/rust-clippy#14880]
### Summary

>I didn't find a corresponding issue before creating this, so sorry if this bug was already known.

Here is an example `clippy.toml` file:
```toml
disallowed-types = [
  { path = "core::ffi::c_int", replacement = "std::ffi::c_int" },
]
```
Even though in a reproducer example I use `std::ffi::c_int` (which is a replacement type, though correct one), it still creates a warning on a use of disallowed type.

I think, that is because of chained re-exports of each in `std`: `core::ffi::c_int` <-> `std::ffi::c_int`. Also, there is no such problem if we use `std::os::raw::c_int` (it seems it is not used in this re-export chain).

### Lint Name

disallowed-types

### Reproducer

I tried this code:

```rust
fn something() -> std::ffi::c_int { todo!() }
```

I saw this happen:

```bash
$ cargo clippy
warning: use of a disallowed type `core::ffi::c_int`
 --> src/main.rs:1:23
  |
1 | pub fn something() -> std::ffi::c_int { todo!() }
  |                       ^^^^^^^^^^^^^^^
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#disallowed_types
  = note: `#[warn(clippy::disallowed_types)]` on by default
```

I expected to see this happen:
No warning should have happened.

### Version

```text
rustc 1.86.0 (05f9846f8 2025-03-31)
binary: rustc
commit-hash: 05f9846f893b09a1be1fc8560e33fc3c815cfecb
commit-date: 2025-03-31
host: x86_64-unknown-linux-gnu
release: 1.86.0
LLVM version: 19.1.7
```

**Comment by blyxyas** [comment]
Does this also happen with other re-exported types or only from `std` <-> `core`?

**Comment by fidelicura** [comment]
> Does this also happen with other re-exported types or only from `std` <-> `core`?

Yes. For example, the same happens within `std` <-> `alloc`:
```
warning: use of a disallowed type `alloc::vec::Vec`
 --> src/main.rs:7:1
  |
7 | use std::vec::Vec;
  | ^^^^^^^^^^^^^^^^^^ help: use: `std::vec::Vec`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#disallowed_types
  = note: `#[warn(clippy::disallowed_types)]` on by default
```
or within `std` <-> `std`:
```
warning: use of a disallowed type `std::collections::hash_map::HashMap`
 --> src/main.rs:1:1
  |
1 | use std::collections::HashMap;
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: use: `std::collections::HashMap`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#disallowed_types
  = note: `#[warn(clippy::disallowed_types)]` on by default
```
So, that's probably on any sort of re-exported types.

Also, when I tried this on re-exported extension traits, I got the following results:
```
$ cargo clippy
    Checking clippy_lint_bug v0.1.0 (/home/kbezuglyi/Desktop/garbage/clippy_lint_bug)
warning: use of a disallowed type `clippy_lint_bug::ext::MyExt`
  --> src/main.rs:14:5
   |
14 |     pub use crate::ext::MyExt as MyExtReexport;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: use: `clippy_lint_bug::ext::MyExtReexport`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#disallowed_types
   = note: `#[warn(clippy::disallowed_types)]` on by default

warning: use of a disallowed type `clippy_lint_bug::ext::MyExt`
  --> src/main.rs:17:1
   |
17 | use ext::MyExtReexport;
   | ^^^^^^^^^^^^^^^^^^^^^^^ help: use: `clippy_lint_bug::ext::MyExtReexport`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#disallowed_types

warning: `clippy_lint_bug` (bin "clippy_lint_bug") generated 2 warnings (run `cargo clippy --fix --bin "clippy_lint_bug"` to apply 2 suggestions)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.06s
```
It at least shows the re-export chain location, but that's maybe because I introduced this directly, not just in some separate foreign crate.

**Comment by Jarcho** [comment]
This is very much intentional. `core::ffi::c_int` is exactly the same type as `std::ffi::c_int` since it's just a re-export.

A separate `disallowed_path` lint could be added, but it could only really work for fully qualified paths. e.g. something like the following wouldn't lint:
```rust
use core::ffi;
fn something() -> ffi::c_int { todo!() }
```

**Comment by fidelicura** [comment]
> This is very much intentional. `core::ffi::c_int` is exactly the same type as `std::ffi::c_int` since it's just a re-export.

I understand they're identical types (`core::ffi::c_int` is indeed a type alias for the same underlying type as `std::ffi::c_int`). However, in a large codebase, having inconsistent usage (some files using `std::ffi::c_int` and others using `core::ffi::c_int`, or even `libc::c_int`) can create unnecessary confusion about which one to use, even though they're functionally equivalent.

The original goal of this change was to standardize on one consistent way to import these C types throughout the codebase, eliminating any potential for confusion in places where the choice should be unambiguous. Also, making it a compile warning will be much better, than make a statement about that usage in something like `CONTRIBUTING.md`.

**Comment by blyxyas** [comment]
> However, in a large codebase, having inconsistent usage (some files using std::ffi::c_int and others using core::ffi::c_int, or even libc::c_int) can create unnecessary confusion about which one to use

Maybe rustfmt could do something about this, adding something like `replaced_paths` to their configuration file. Otherwise, adding a pre-commit test with a shell script to check for grep for `core::ffi::c_int` and your other disallowed paths could be the way to go.
